Roman/azure cognitive embeddings

rbiseck3 opened the pull request:
### Description
This PR is two-fold:  

**Embeddings:**
* Embeddings incorporated into the sharepoint source connector, which will now call out to OpenAI and create embeddings if the flag is passed in and the api key provided. 

**Writing vector content (embeddings) to Azure cognitive search index:**
* The schema for the index expected to exist in Azure has been updated to include the vector field type and a test script has been added to test the new content being produced from the Sharepoint connector to push the embedding content.  

Some important notes about other changes in here:
* The embedding code had to be updated to patch the `to_dict` method on elements to add `embeddings` to the dict output if that was added. While the code originally added the embedding content, when `to_dict` was called to save the content as json, this was lost.

ryannikolaidis commented on CHANGELOG.md:
```diff
@@ -1,11 +1,12 @@
 * **Add functionality to save embedded images in PDF's separately as images** This allows users to save embedded images in PDF's separately as images, given some directory path. The saved image path is written to the metadata for the Image element. Downstream applications may benefit by providing users with image links from relevant "hits."
 * **Azure Cognite Search destination connector** New Azure Cognitive Search destination connector added to ingest CLI.  Users may now use `unstructured-ingest` to write partitioned data from over 20 data sources (so far) to an Azure Cognitive Search index.
 * **Improves salesforce partitioning** Partitions Salesforce data as xlm instead of text for improved detail and flexibility. Partitions htmlbody instead of textbody for Salesforce emails. Importance: Allows all Salesforce fields to be ingested and gives Salesforce emails more detailed partitioning.
+* **Embeddings support in ingest CLI** The sharepoint connector can now optionally create embeddings from the elements it pulls out during partition and upload those embeddings to Azure Cognitive Search index.
```
```suggestion
* **Embeddings support for the SharePoint SourceConnector via unstructured-ingest CLI** The SharePoint connector can now optionally create embeddings from the elements it pulls out during partition and upload those embeddings to Azure Cognitive Search index.
```

ryannikolaidis commented on test_unstructured_ingest/test-ingest-azure-cognitive-search.sh:
```diff
@@ -4,11 +4,13 @@ set -e
 SCRIPT_DIR=$(dirname "$(realpath "$0")")
 cd "$SCRIPT_DIR"/.. || exit 1
-OUTPUT_FOLDER_NAME=s3
+OUTPUT_FOLDER_NAME=s3-azure-dest
```
```suggestion
OUTPUT_FOLDER_NAME=s3-azure_cog_search-dest
```
?

...since we have a few different azure products

ryannikolaidis commented on test_unstructured_ingest/test-ingest-azure-cognitive-search.sh:
```diff
@@ -65,16 +67,27 @@ PYTHONPATH=. ./unstructured/ingest/main.py \
+
+docs_count=0
+attempt=1
+while [ "$docs_count" -eq 0 ] && [ "$attempt" -lt 6 ]; do
```
quick comment above on what we're doing here.

ryannikolaidis commented on test_unstructured_ingest/test-ingest-sharepoint-embed-cog-index.sh:
```diff
@@ -0,0 +1,110 @@
+echo "Creating index $DESTINATION_INDEX"
+response_code=$(curl -s -o /dev/null -w "%{http_code}" -X PUT \
+"https://utic-test-ingest-fixtures.search.windows.net/indexes/$DESTINATION_INDEX?api-version=$API_VERSION" \
+--header "api-key: JV1LDVRivKEY9J9rHBQqQeTvaGoYbD670RWRaANxaTAzSeDy8Eon" \
```
don't hardcode api key here.

ryannikolaidis commented on test_unstructured_ingest/test-ingest-azure-cognitive-search.sh:
```diff
@@ -65,16 +67,27 @@ PYTHONPATH=. ./unstructured/ingest/main.py \
+
+  # Check the contents of the index
+  docs_count=$(curl "https://utic-test-ingest-fixtures.search.windows.net/indexes/$DESTINATION_INDEX/docs/\$count?api-version=$API_VERSION" \
+    --header "api-key: $AZURE_SEARCH_API_KEY" \
+    --header 'content-type: application/json' | jq)
+
+  echo "docs count pulled from Azure: $docs_count"
+
+  attempt=$((attempt+1))
+done
+
 
 expected_docs_count=0
-for i in $(jq length "$OUTPUT_DIR"/*); do
+for i in $(jq length "$OUTPUT_DIR"/**/*.json); do
   expected_docs_count=$((expected_docs_count+i));
```
maybe we update the names of "docs_count" vs "expected_docs_count"?... since we have a lot going on here maybe something to make a name to make it clear that docs_count are the local written results and expected_docs_count is what was written to the destination?

ryannikolaidis commented on test_unstructured_ingest/test-ingest-sharepoint-embed-cog-index.sh:
```diff
@@ -0,0 +1,110 @@
+    --endpoint "$AZURE_SEARCH_ENDPOINT" \
+    --index "$DESTINATION_INDEX"
+
+docs_count=0
+attempt=1
+while [ "$docs_count" -eq 0 ] && [ "$attempt" -lt 6 ]; do
+  echo "attempt $attempt: sleeping 10 seconds to let index finish catching up after writes"
+  sleep 10
+
+  # Check the contents of the index
+  docs_count=$(curl "https://utic-test-ingest-fixtures.search.windows.net/indexes/$DESTINATION_INDEX/docs/\$count?api-version=$API_VERSION" \
+    --header "api-key: $AZURE_SEARCH_API_KEY" \
+    --header 'content-type: application/json' | jq)
+
+  echo "docs count pulled from Azure: $docs_count"
+
+  attempt=$((attempt+1))
+done
+
+
+expected_docs_count=0
+for i in $(jq length "$OUTPUT_DIR"/**/*.json); do
+  expected_docs_count=$((expected_docs_count+i));
+done
+
+
+if [ "$docs_count" -ne "$expected_docs_count" ];then
+  echo "Number of docs $docs_count doesn't match the expected docs: $expected_docs_count"
+  exit 1
+fi
```
same as notes above

ryannikolaidis commented on unstructured/ingest/cli/interfaces.py:
```diff
@@ -184,3 +185,40 @@ def add_cli_options(cmd: click.Command) -> None:
+        options = [
+            click.Option(
+                ["--embedding-api-key"],
+                help="openai api key",
+                envvar="OPENAI_API_KEY",
```
probably shouldn't hardcode openai naming assumption here, since in the future will be relevant to the chosen model / location

ryannikolaidis commented on unstructured/ingest/cli/interfaces.py:
```diff
@@ -184,3 +185,40 @@ def add_cli_options(cmd: click.Command) -> None:
+        *,
+        infer_missing=False,
+    ):
+        if isinstance(kvs, dict):
```
docstring explaining what this is and why needed

ryannikolaidis commented on unstructured/ingest/runner/sharepoint.py:
```diff
@@ -1,69 +1,66 @@
-    **kwargs,
-):
-    writer_kwargs = writer_kwargs if writer_kwargs else {}
+class Sharepoint(Runner):
```
```suggestion
class SharePoint(Runner):
```
for consistency with naming [here](https://www.microsoft.com/en-us/microsoft-365/sharepoint/collaboration)?

rbiseck3 commented on unstructured/staging/base.py:
```diff
@@ -41,7 +41,7 @@ def _get_metadata_table_fieldnames():
 def convert_to_isd(elements: List[Element]) -> List[Dict[str, Any]]:
     """Represents the document elements as an Initial Structured Document (ISD)."""
-    isd: List[Dict[str, str]] = []
+    isd: List[Dict[str, Any]] = []
```
The return type is set to `List[Dict[str, Any]]`:[base.py#L42](https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/staging/base.py#L42). So this just helps it match that:
```python
def convert_to_isd(elements: List[Element]) -> List[Dict[str, Any]]:
...
```

ryannikolaidis replied:
yes, I understand, that was the part that I appreciated, but if we're bothering to update this unrelated change wondering if we can be more specific here.

ryannikolaidis replied:
probably fine, looks like we're doing this everywhere.

ryannikolaidis commented on CHANGELOG.md:
```diff
@@ -1,12 +1,12 @@
 * **Azure Cognite Search destination connector** New Azure Cognitive Search destination connector added to ingest CLI.  Users may now use `unstructured-ingest` to write partitioned data from over 20 data sources (so far) to an Azure Cognitive Search index.
 * **Improves salesforce partitioning** Partitions Salesforce data as xlm instead of text for improved detail and flexibility. Partitions htmlbody instead of textbody for Salesforce emails. Importance: Allows all Salesforce fields to be ingested and gives Salesforce emails more detailed partitioning.
-
+* **Embeddings support for the SharePoint SourceConnector via unstructured-ingest CLI** The SharePoint connector can now optionally create embeddings from the elements it pulls out during partition and upload those embeddings to Azure Cognitive Search index.
```
missed earlier: should also include "why this matters"

ryannikolaidis commented on test_unstructured_ingest/test-ingest-sharepoint-embed-cog-index.sh:
```diff
@@ -0,0 +1,113 @@
+  fi
+}
+
+#trap cleanup EXIT
```
no cleanup? intentional?

ryannikolaidis commented on unstructured/ingest/interfaces.py:
```diff
@@ -61,6 +63,20 @@ class ReadConfig(BaseConfig):
+        }
+        if self.model_name:
+            kwargs["model_name"] = self.model_name
+        return OpenAIEmbeddingEncoder(**kwargs)
```
this always returns an instance of OpenAIEmbeddingEncoder? would have assumed the encoder was determined by a parameter? or is this placeholder until we have more than one? (if so, probably good to mark a TODO on this)

ryannikolaidis commented:
just a few other small notes, otherwise this looks good! 

question: should we have `test-ingest-sharepoint-embed-cog-index.sh` at least run on pushes to main? don't think we need to burn through credits on every push to ever pr branch, but having it check / keep us honest every once and a while would be good. btw: I just added OPENAI_API_KEY secret to the repo.

ryannikolaidis commented:
exported secrets and tried running the test manually with:

```
PYTHONPATH=. ./test_unstructured_ingest/test-ingest-sharepoint-embed-cog-index.sh
```

but getting this error:

```
Creating index utic-test-ingest-fixtures-output-1695694952
Index creation success: 201
Usage: main.py sharepoint azure-cognitive-search [OPTIONS]
Try 'main.py sharepoint azure-cognitive-search --help' for help.

Error: No such option: --embedding-api-key
```

ryannikolaidis commented:
yea, not seeing the embedding options with:

```
PYTHONPATH=. ./unstructured/ingest/main.py sharepoint azure-cognitive-search --help
Usage: main.py sharepoint azure-cognitive-search [OPTIONS]

Options:
  --key TEXT       Key credential used for authenticating to an Azure service.
                   [env var: AZURE_SEARCH_API_KEY; required]
  --endpoint TEXT  The URL endpoint of an Azure search service. In the form of
                   https://{{service_name}}.search.windows.net  [env var:
                   AZURE_SEARCH_ENDPOINT; required]
  --index TEXT     The name of the index to connect to  [required]
  --help           Show this message and exit.
```

rbiseck3 commented:
@ryannikolaidis, sorry, when I removed support for the env var to populate the CLI arg, I added it to the script but put it for the dest rather than the source connector. This has all been fixed now.

rbiseck3 commented:
Follow up PR to come that will introduce a CI job to run only on `main` to test the embedding since this hits the openai api and we don't want to use up the api credits too fast.

ryannikolaidis commented:
requires pip install of tiktoken:

```
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/error.py", line 22, in wrapper
    raise cls(cls.error_string.format(str(error))) from error
unstructured.ingest.error.EmbeddingEncoderConnectionError: Error in connecting to the embedding model provider: Could not import tiktoken python package. This is needed in order to for OpenAIEmbeddings. Please install it with `pip install tiktoken`.
```

we should probably extend as an extra included with  langchain and openai which are needed for embedding[ here](https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/embed/openai.py#L42-L57)

ryannikolaidis reviewed (approved):
lgtm! thanks for the changes here.

ryannikolaidis commented:
update title and description to clarify this is only for sharepoint-cognitive search?
